File: python/packages/autogen-ext/src/autogen_ext/task/_source_match_termination.py
from typing import Sequence, List

from autogen_agentchat.base import TerminationCondition, TerminatedException
from autogen_agentchat.messages import StopMessage, AgentMessage


class SourceMatchTermination(TerminationCondition):
    """Terminate the conversation after a specific source responds.

    Args:
        sources (List[str]): List of source names to terminate the conversation.

    Raises:
        TerminatedException: If the termination condition has already been reached.
    """

    def __init__(self, sources: List[str]) -> None:
        self._sources = sources
        self._terminated = False

    @property
    def terminated(self) -> bool:
        return self._terminated

    async def __call__(self, messages: Sequence[AgentMessage]) -> StopMessage | None:
        if self._terminated:
            raise TerminatedException("Termination condition has already been reached")
        if not messages:
            return None
        last_message = messages[-1]
        if last_message.source in self._sources:
            self._terminated = True
            return StopMessage(content=f"'{last_message.source}' answered", source="SourceMatchTermination")
        return None

    async def reset(self) -> None:
        self._terminated = False

File: dotnet/src/Microsoft.AutoGen/AgentChat/Terminations/SourceMatchTermination.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// SourceMatchTermination.cs

using Microsoft.AutoGen.AgentChat.Abstractions;

namespace Microsoft.AutoGen.AgentChat.Terminations;

/// <summary>
/// Terminate the conversation after a specific source responds.
/// </summary>
public sealed class SourceMatchTermination : ITerminationCondition
{
    /// <summary>
    /// Initializes a new instance of the <see cref="SourceMatchTermination"/> class.
    /// </summary>
    /// <param name="sources">List of source names to terminate the conversation.</param>
    public SourceMatchTermination(params IEnumerable<string> sources)
    {
        this.Sources = [.. sources];
    }

    public HashSet<string> Sources { get; }
    public bool IsTerminated { get; private set; }

    public ValueTask<StopMessage?> CheckAndUpdateAsync(IList<AgentMessage> messages)
    {
        if (this.IsTerminated)
        {
            throw new TerminatedException();
        }

        foreach (AgentMessage item in messages)
        {
            if (this.Sources.Contains(item.Source))
            {
                this.IsTerminated = true;
                string message = $"'{item.Source}' answered.";
                StopMessage result = new() { Content = message, Source = nameof(SourceMatchTermination) };
                return ValueTask.FromResult<StopMessage?>(result);
            }
        }

        return ValueTask.FromResult<StopMessage?>(null);
    }

    public void Reset()
    {
        this.IsTerminated = false;
    }
}
